libr/core/cmd_search.c

```diff
@@ -1549,6 +1550,141 @@ static void do_esil_search(RCore *core, struct search_parameters *param, const c
+	RListIter *iter;
+	const int mininstrsz = r_anal_archinfo (core->anal, R_ANAL_ARCHINFO_MIN_OP_SIZE);
+        const int minopcode = R_MAX (1, mininstrsz);
+	int stacksize = r_config_get_i (core->config, "esil.stack.depth");
+	int iotrap = r_config_get_i (core->config, "esil.iotrap");
+	const char *a0 = r_reg_get_name (core->anal->reg, R_REG_NAME_SN);
+	const char *pc = r_reg_get_name (core->dbg->reg, R_REG_NAME_PC);
+	struct r_reg_item_t *r = r_reg_get (core->dbg->reg, pc, -1);
+	struct r_reg_item_t *reg_a0 = r_reg_get (core->dbg->reg, a0, -1);
+	
+	if (!(esil = r_anal_esil_new (stacksize, iotrap))) {
+		return;
+	}
+	memset (previnstr, 0, sizeof (previnstr));
+	buf = malloc (bsize);
+	arr = malloc (bsize);
+	if (!buf || !arr) {
+		eprintf ("Cannot allocate %d bytes\n", bsize);
+		free(buf);
+		free(arr);
+		return;
+	}
+	r_cons_break_push (NULL, NULL);
+	r_list_foreach (param->boundaries, iter, map) {
+		ut64 from = map->itv.addr;
+		ut64 to = r_itv_end (map->itv);
+		for (i = 0, at = from; at < to; at++, i++) {
+			if (r_cons_is_breaked ()) {
+				break;
+			}
+			if (i >= (bsize - 32)) {
+				i = 0;
+			}
+			if (!i) {
+				r_core_read_at (core, at, buf, bsize);
+			}
+			ret = r_anal_op (core->anal, &aop, at, buf + i, bsize - i);
+			curpos = idx++ % (MAXINSTR + 1);
+			previnstr[curpos] = ret; // This array holds prev n instr size + cur instr size
+			if (ret) {
+				bool match = false;
+				if (aop.type == R_ANAL_OP_TYPE_SWI) {
+					match = true;
+				}
+				if (match) {
+					// This for calculating no of bytes to be subtracted , to get n instr above syscall
+					int nbytes = 0;
+					int nb_opcodes = MAXINSTR;
+					SUMARRAY (previnstr, nb_opcodes, nbytes);
+					curpc = at - (nbytes - previnstr[curpos]);
+					r_reg_set_value (core->dbg->reg, r, curpc);
+					// This loop is for emulating n instr above the syscall
+					for (j = 0; curpc < at; curpc++, j++) {
+						if (j >= (bsize - 32)) {
+							j = 0;
+						}
+						if (!j) {
+							r_core_read_at (core, curpc, arr, bsize);
+						}
+						inslen = r_anal_op (core->anal, &aop, curpc, arr + j, bsize - j);
+						if (inslen) {
+							bool tmatch = false;
+							if (isnonlinear (aop.type)) {
+								tmatch = true;	
+							}
+							int incr = (core->search->align > 0)? core->search->align - 1:  inslen - 1;
+							if (incr < 0) {
+								incr = minopcode;
+							}	
+							j += incr;
+							curpc += incr;
+							if (tmatch) {	// skip the instr
+								r_reg_set_value (core->dbg->reg, r, curpc + 1);
+							} else {	// step instr
+								r_core_esil_step (core, UT64_MAX, NULL, NULL);
+							}
+						}
+					}
+					int off = r_debug_reg_get (core->dbg, a0);
+					RSyscallItem *item = r_syscall_get (core->anal->syscall, off, -1);
+					if (item) {
+						r_cons_printf ("0x%08"PFMT64x" %s\n", at, item->name);	
+					}
+					memset (previnstr, 0, sizeof (previnstr)); // clearing the buffer
+					r_reg_set_value (core->dbg->reg, reg_a0, -2); // clearing register A0
```

Comment:
a0 = -2 ? why -2?

Reply:
Because for some reason syscall api with param -1 takes that as 0 and returns a valid syscall function , hence choosing -2

Reply:
try ```as -1``` with any arch